dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,514 @@
+Multiboot2 support could be easily added.
+
+#### Fobnail integration
+
+Running [fobnail-attester](https://github.com/fobnail/fobnail-attester) on LK
+would be problematic to say the last:
+
+- LK lacks most drivers including USB host drivers (UHCI, OHCI, EHCI, XHCI),
+  USB EEM driver and TPM drivers.
+
+- Lack of cryptographic library in LK. TLK does provide one (potentially could be).
+
+- Default C library is too much constrained - provides only most basic API like
+  string operations, printing, etc. Lacks anything more complex, including
+  networking support. There is a LK's
+  [fork](https://github.com/littlekernel/newlib) of Newlib, however of unknown
+  quality.
+
+- TCP/IP is provided by `minip` library (which is part of LK). Attester depends
+  on `libcoap3` which either would have to be ported to `minip` or `minip` would
+  have to be integrated with `newlib`.
+
+- LK has no suitable bootloader capabilities.
+  [lkboot](https://github.com/littlekernel/lk/tree/master/app/lkboot) could
+  serve as reference to develop custom bootloader.
+
+### Fuchsia (Zircon)
+
+[Fuchsia](https://fuchsia.dev) is a general-purpose OS developed by Google. Its
+purpose is to replace GNU/Linux in Google products (like Android, Chrome OS)
+with Zircon microkernel (used to be known as Magenta) and custom userland.
+Fuchsia runs on x86-64 and ARM64. There used to be an
+[unofficial](https://github.com/slavaim/riscv-magenta) RISC-V port, however it
+has been unmaintained for years. Support for 32-bit architectures was dropped
+some time ago.
+
+Zircon is based on LK, however similarities between these two are small and
+constantly decreasing. LK is a kernel intended for embedded usage, while Zircon
+is a general-purpose microkernel that could compete with Linux.
+
+#### Running in DLME
+
+Historically Zircon used Multiboot 1 for booting. Now it uses ZBI with a UEFI
+bootloader called Gigaboot. Multiboot 1 support is still present, but SKL
+doesn't support it anyway. See [Zephyr](#running-in-dlme) section above for
+details.
+
+#### Fobnail integration
+
+Fuchsia provides a Unix-like environment with musl C library. Network is
+available but is no EEM driver, and only XHCI host driver is supported. Since
+EHCI is still widely used, this is a major limitation. TPM driver is available
+but without `libtss2`.
+
+Zircon has a kexec-like capability which allows to boot another Zircon instance.
+This is known as `mexec`.
+
+Following things would have to be done:
+
+- Bring `libtss2` stack to Zircon.
+
+- Bring EEM driver. There is ECM driver available, which could serve as
+  reference.
+
+- Bring at least USB EHCI driver, ideally should support UHCI and OHCI too.
+
+- Figure out how to use `mexec` to boot Linux.
+
+## PoC test
+
+Running DLME requires GRUB from TrenchBoot as mainline doesn't have DLME
+support. Currently, there is an ongoing discussion how TrenchBoot should be
+integrated into Linux. Unless this is solved no investment in TrenchBoot GRUB2
```

Comment:
@arturkow2000 please add link to that discussion or couple links, I'm trying to find references but on LKML comes to mind. There was definitely something wrote by Mattew Garret. @krystian-hebel do you know where to find most recent status of above discussion?

Reply:
https://groups.google.com/g/trenchboot-devel should have copies of all of public discussion in one place. If LKML link is preferred, it should be easy to find using subjects from `trenchboot-devel` list.

Reply:
Added 3 links